### PR TITLE
Boto s3 bucket state

### PR DESCRIPTION
**Pull request by kraney**
This module is for managing S3 buckets - not the content, but the configuration of the bucket itself. Related: #6443

This allows an S3 bucket to be created specifying location, ACLs, CORS, lifecycle, logging, notifications, policy, replication, payment, tagging, versioning, and website config. Module and state module are both included.

Note that S3 has a few quirks to be aware of. If no IDs are specified for items in your lifecycle, policy, and replication configurations, this is accepted and a random unique ID is generated. This means that these config items will never match the specified configuration when executing "present", so they'll be updated every time. That's probably harmless, but possibly annoying. Avoid it by specifying your own IDs rather than allowing them to be autogenerated.

Also, there doesn't seem to be any way to determine canonical user IDs based on the information accepted for setting them when configuring ACLs other than setting the ACL and seeing what it looks like when it comes back. Thus ACLs may also always update. Avoid by setting the ACL, using describe to see the result, and base your config on that directly using the "AccessControlPolicy" option for setting ACL. Again, updating an ACL to a completely equivalent setting is probably harmless in any case. It just clutters the logs.

### CODE REVIEW

File: salt/modules/boto_s3_bucket.py [hunk 1]
```diff
@@ -0,0 +1,831 @@
...
+                [{...}] \\
+                [{...}]
+
+    '''
+
+    try:
+        conn = _get_conn(region=region, key=key, keyid=keyid, profile=profile)
+        if TopicConfigurations is None:
+            TopicConfigurations = []
+        elif isinstance(TopicConfigurations, string_types):
```
Comment (cachedout) [inline salt/modules/boto_s3_bucket.py]: Let's use `six.string_types` in these cases for forward-compatibility.

Comment (kraney) [inline salt/modules/boto_s3_bucket.py]: It is using six.string_types. Do you mean for me to change the import so it's literally six.string_types?

**Comment by kraney** [comment]
I'm afraid I don't understand the lint error shown. How can a blank line be a Python 3 problem?

**Comment by cachedout** [comment]
@kraney Looks like you figured it out. I wish that Py3 test displayed that error more sensibly. Thanks for digging in and figuring it out. :]

File: salt/modules/boto_s3_bucket.py (hunk 1 above)
Comment (cachedout) [inline salt/modules/boto_s3_bucket.py]: Whoops! That is my bad. You're right. Thanks.

**Pull request merged**